invoiceregistry_test.go

```diff
@@ -186,6 +186,100 @@ func (r *registryTestContext) checkHtlcSettled(resolution HtlcResolution) {
 	require.IsType(r.t, &HtlcSettleResolution{}, resolution)
 }
 
+// optNotifyHtlc is a functional option to notifyHtlc.
+type optNotifyHtlc func(h *registryHtlc)
+
+// optRHash sets the rHash in the registryHtlc.
+func optRHash(hash lntypes.Hash) optNotifyHtlc {
+	return func(h *registryHtlc) {
+		h.rHash = hash
+	}
+}
+
+// optResolveSend sends the HTLC resolution over the provided channel.
+func optResolveSend(resolved chan<- HtlcResolution) optNotifyHtlc {
+	return func(h *registryHtlc) {
+		h.resolve = func(r HtlcResolution) {
+			resolved <- r
+		}
+	}
+}
+
+// optResolveClose closes the channel when an HTLC resolution is received.
+func optResolveClose(resolved chan<- struct{}) optNotifyHtlc {
+	return func(h *registryHtlc) {
+		h.resolve = func(r HtlcResolution) {
+			close(resolved)
+		}
+	}
+}
+
+// optAmtPaid changes the amtPaid field of the HTLC.
+func optAmtPaid(amtPaid int64) optNotifyHtlc {
+	return func(h *registryHtlc) {
+		h.amtPaid = msat(amtPaid)
+	}
+}
+
+// optPayAddr creates an MPP payload if necessary and sets its payAddr. The
+// amt is copied from the HTLC's amtPaid field if it's not already set for the
+// payload.
+func optPayAddr(payAddr [32]byte) optNotifyHtlc {
+	return func(h *registryHtlc) {
+		if _, ok := h.payload.(*testPayload); !ok {
+			h.payload = &testPayload{
+				amt: h.amtPaid,
+			}
+		}
+
+		h.payload.(*testPayload).payAddr = payAddr
+	}
+}
+
+// optPayloadAmt creates an MPP payload if necessary and sets its amt. The
+// payAddr is left empty if it's not already set for the payload.
+func optPayloadAmt(amt int64) optNotifyHtlc {
+	return func(h *registryHtlc) {
+		if _, ok := h.payload.(*testPayload); !ok {
+			h.payload = &testPayload{}
+		}
+
+		h.payload.(*testPayload).amt = msat(amt)
+	}
+}
+
+// optHtlcID sets the HTLC ID for the HTLC's circuit key.
+func optHtlcID(id uint64) optNotifyHtlc {
+	return func(h *registryHtlc) {
+		h.circuitKey.HtlcID = id
+	}
+}
+
+// optCurrentHeight sets the HTLC's current block height.
+func optCurrentHeight(height int32) optNotifyHtlc {
+	return func(h *registryHtlc) {
+		h.currentHeight = height
+	}
+}
+
+// notifyHtlc notifies the registry of an incoming HTLC. It builds the HTLC
+// with convenient defaults but requires a minimum of an optRHash and an
```

Comment:
Done. Since there's not a `nil` for the `[32]byte` type of the `payAddr`, I've also added a `payAddrZero` var to compare/assign when intentionally attempting to create an HTLC without an MPP payload.

Reply:
Doesn't that mean that `payAddr` isn't a required parameter after all?

Reply:
It's required by `process`, but that requirement is part of what we test in the new `TestMPPValidity`.

Reply:
We could remove the `payAddrZero` and check, and instead create an `optNilPayload` that sets the payload to `(*nullPayload)(nil)` in the option instead?

Reply:
Agreed